include/albatross/src/covariance_functions/covariance_function.hpp

```diff
@@ -135,15 +137,71 @@ template <typename CovFunc, typename Caller> struct MeasurementForwarder {
   }
 };
 
+template <typename CovFunc, typename Caller> struct VariantForwarder {
+
+  template <
+      typename X, typename Y,
+      typename std::enable_if<
+          has_valid_cov_caller<CovFunc, Caller, X, Y>::value, int>::type = 0>
+  static double call(const CovFunc &cov_func, const X &x, const Y &y) {
+    return Caller::call(cov_func, x, y);
+  }
+
+  template <
+      typename A, typename B, typename C,
+      typename std::enable_if<
+          !has_valid_cov_caller<CovFunc, Caller, variant<A, B>, C>::value &&
+              has_valid_cov_caller<CovFunc, Caller, A, C>::value &&
+              has_valid_cov_caller<CovFunc, Caller, B, C>::value,
+          int>::type = 0>
+  static double call(const CovFunc &cov_func, const variant<A, B> &x,
+                     const C &y) {
+
+    return x.match([&y, &cov_func](A xx) { return call(cov_func, xx, y); },
```

Comment:
Have you tried having these lambdas take a parameter of `const A&`? That would avoid a copy of `A`.

Reply:
Nice!  That works. I tried it before without luck ... but must have been a different issue.